goldenmaster/apigear/mqtt/mqttcppclient.cpp

```diff
@@ -0,0 +1,573 @@
+                m_invokeResults.erase(randomId);
+            }
+            m_invokeResultsMutex.unlock();
+            if(func) {
+                const InvokeReplyArg arg{nlohmann::json::parse(message.content)};
+                func(arg);
+            }
+        }
+    }
+}
+
+void Client::invokeRemote(const Topic& topic, const std::string& value, InvokeReplyFunc func)
+{
+    const std::string responseTopic = topic.getEncodedTopic() + "/" + m_clientID + "/result" ;
+
+    MQTTAsync_responseOptions opts = MQTTAsync_responseOptions_initializer;
+    MQTTAsync_message pubmsg = MQTTAsync_message_initializer;
+    int rc = 0;
+
+    MQTTProperty responseTopicProperty;
+    responseTopicProperty.identifier = MQTTPROPERTY_CODE_RESPONSE_TOPIC;
+    responseTopicProperty.value.data = { static_cast<int>(responseTopic.size()), const_cast<char*>(responseTopic.c_str()) };
+    MQTTProperties_add(&(opts.properties), &responseTopicProperty);
+
+    MQTTProperty correlationDataProperty;
+    correlationDataProperty.identifier = MQTTPROPERTY_CODE_CORRELATION_DATA;
+    std::uniform_int_distribution<> distribution (0, 100000);
+
+    auto randomId = 0;
+    m_invokeResultsMutex.lock();
+    do {
+        randomId = distribution(randomNumberGenerator);
+    } while (m_invokeResults.find(randomId) != m_invokeResults.end());
+    m_invokeResults.insert(std::pair<int, InvokeReplyFunc>(randomId, func));
+    m_invokeResultsMutex.unlock();
+    const std::string correlationData = { std::to_string(randomId) } ;
+    correlationDataProperty.value.data = { static_cast<int>(correlationData.size()), const_cast<char*>(correlationData.c_str()) };
+    MQTTProperties_add(&(opts.properties), &correlationDataProperty);
+
+    opts.onSuccess5 = onSendSuccess;
+    opts.onFailure5 = onSendFailure;
+    opts.context = this;
+    // opts.properties = properties;
+    pubmsg.payload = const_cast<void*>(static_cast<const void*>(value.c_str()));
+    pubmsg.payloadlen = value.size();
+    pubmsg.qos = QOS;
+    pubmsg.retained = 0;
+
+    // the responseOptions properties do get overwritten by the msg properties later
+    pubmsg.properties = opts.properties;
+
+    if ((rc = MQTTAsync_sendMessage(*m_client.get(), topic.getEncodedTopic().c_str(), &pubmsg, &opts)) != MQTTASYNC_SUCCESS)
+    {
+        AG_LOG_ERROR("Failed to start sendMessage, return code " + std::to_string(rc));
+    }
+}
+
+void Client::notifyPropertyChange(const Topic& topic, const std::string& value)
+{
+    MQTTAsync_responseOptions opts = MQTTAsync_responseOptions_initializer;
+    MQTTAsync_message pubmsg = MQTTAsync_message_initializer;
+    int rc = 0;
+
+    opts.onSuccess5 = onSendSuccess;
+    opts.onFailure5 = onSendFailure;
+    opts.context = this;
+    pubmsg.payload = const_cast<void*>(static_cast<const void*>(value.c_str()));
+    pubmsg.payloadlen = value.size();
+    pubmsg.qos = QOS;
+    // property changes shall be retained and automatically send to new clients
+    pubmsg.retained = 1;
```

Comment:
Only the last one per topic is retained. If a new client connects, it automatically gets only the latest state/msg sent.

Reply:
maybe change a comment? that last one for topic is retained? or that is common knowledge and it was just me who needed explanation 😅

Reply:
I think when you are working with MQTT and in this specific function, setting a property via topic it should be clear. Otherwise let me know and we can more explanation or refer to some mqtt docs